Bottom: d584b4b1c2416f18c5564bd31225cba9138d84a9
Top:    fb19f003dc1f917d4ad2e32c9b2db10378e75042
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-09-14 23:27:45 +0100

undo.c: Add error message where RestoreUndoSerialNumber() will break things

If operations are addded to the undo stack between a call to
IncrementUndoSerialNumber() and a subsequent call for
RestoreUndoSerialNumber(), those added operations will be placed in the
undo stack with an inconsistent serial number.

The {Save,Restore}UndoSerialNumber() API is pretty fragile in this
regard, and we should avoid using it where possible.

An better alternative might be to implement a "LockUndoSerialNumber()"
call which temporarily disables IncrementUndoSerialNumber(), and a
"UnlockUndoSerialNumber()" call which re-enables its increment function.

Better yet - we could make these functions nestable, so we need not
worry whether some action uses this new feature internally when we are
also using it.


---

diff --git a/src/undo.c b/src/undo.c
index 7832600..17bebb9 100644
--- a/src/undo.c
+++ b/src/undo.c
@@ -76,6 +76,9 @@
 
 RCSID ("$Id$");
 
+static bool between_increment_and_restore = false;
+static bool added_undo_between_increment_and_restore = false;
+
 /* ---------------------------------------------------------------------------
  * some local data types
  */
@@ -254,6 +257,9 @@ GetUndoSlot (int CommandType, int ID, int Kind)
 	break;
       }
 
+  if (between_increment_and_restore)
+    added_undo_between_increment_and_restore = true;
+
   /* copy typefield and serial number to the list */
   ptr = &UndoList[UndoN++];
   ptr->Type = CommandType;
@@ -1177,6 +1183,10 @@ Redo (bool draw)
 void
 RestoreUndoSerialNumber (void)
 {
+  if (added_undo_between_increment_and_restore)
+    Message (_("ERROR: Operations were added to the Undo stack with an incorrect serial number\n"));
+  between_increment_and_restore = false;
+  added_undo_between_increment_and_restore = false;
   Serial = SavedSerial;
 }
 
@@ -1187,6 +1197,8 @@ void
 SaveUndoSerialNumber (void)
 {
   Bumped = false;
+  between_increment_and_restore = false;
+  added_undo_between_increment_and_restore = false;
   SavedSerial = Serial;
 }
 
@@ -1205,6 +1217,7 @@ IncrementUndoSerialNumber (void)
         SetChangedFlag (true);
       Serial++;
       Bumped = true;
+      between_increment_and_restore = true;
     }
 }
